yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala

```diff
@@ -372,7 +372,14 @@ private[spark] class ApplicationMaster(
             case i: InterruptedException =>
             case e: Throwable => {
               failureCount += 1
-              if (!NonFatal(e) || failureCount >= reporterMaxFailures) {
+              if ("org.apache.hadoop.yarn.exceptions.ApplicationAttemptNotFoundException".equals(
+                e.getClass().getName())) {
```

Comment:
also please add a comment to explain why we need to put this under this case, i.e. this exception was introduced in hadoop 2.x and this code would not compile otherwise